### PR TITLE
Add key vault for users api keys

### PR DESCRIPTION
**Pull request by khaelys**
Provide support for migration to store users' API keys securely in Azure Key Vault.

### CODE REVIEW

**Comment by simone3991** [comment]
closes #35

File: src/Clockify/ClockifySetupDialog.cs [hunk 1]
```diff
@@ -62,15 +66,23 @@ public ClockifySetupDialog(UserState userState, IClockifyService clockifyService
...
             {
                 var userProfile = await _userState.CreateProperty<UserProfile>("UserProfile")
                     .GetAsync(promptContext.Context, () => new UserProfile(), cancellationToken);
 
                 string? userId = _clockifyService.GetCurrentUserAsync(token).Result.Id;
-
-                userProfile.ClockifyToken = token;
+                
+                var tokenData = await _tokenRepository.WriteAsync(token, userProfile.ClockifyTokenId);
+                userProfile.ClockifyToken = tokenData.Value;
```
Comment (simone3991) [inline src/Clockify/ClockifySetupDialog.cs:83]: why is this here? Do we really need a stepped rollout? 
If it's properly tested I think we can go all-in with the KV without having a 2-way phase

Or am I missing something?

File: src/Clockify/ClockifyHandler.cs [hunk 2]
```diff
@@ -125,6 +128,12 @@ public class ClockifyHandler : IBotHandler
             try
             {
                 await _clockifyService.GetCurrentUserAsync(userProfile.ClockifyToken);
+                // It will be removed when only ClockifyTokenId will be used
```
Comment (simone3991) [inline src/Clockify/ClockifyHandler.cs:131]: I thought we'd do something like this, which would automatically put us where we want to be (_all_ tokens present in KV _only_):

```csharp
            switch (userProfile.ClockifyTokenId)
            {
                case null when userProfile.ClockifyToken == null:
                    // first sign in
                    await _clockifySetupDialog.RunAsync(turnContext, _dialogState, cancellationToken);
                    return true;
                case null when userProfile.ClockifyToken != null:
                    // old storage, migrating and cleaning
                    try
                    {
                        await _clockifyService.GetCurrentUserAsync(userProfile.ClockifyToken);
                        var tokenData = await _tokenRepository.WriteAsync(userProfile.ClockifyToken);
                        userProfile.ClockifyToken = null;
                        userProfile.ClockifyTokenId = tokenData.Id;
                        return false;
                    }
                    catch (ErrorResponseException)
                    {
                        // old login invalid, asking again
                        await _clockifySetupDialog.RunAsync(turnContext, _dialogState, cancellationToken);
                        return true;
                    }
                default:
                    try
                    {
                        var token = await _tokenRepository.ReadAsync(userProfile.ClockifyTokenId!);
                        await _clockifyService.GetCurrentUserAsync(token!.Value);
                        userProfile.ClockifyToken = null; // just to be sure :)
                        return false;
                    }
                    catch (ErrorResponseException)
                    {
                        // old login invalid, asking again
                        await _clockifySetupDialog.RunAsync(turnContext, _dialogState, cancellationToken);
                        return true;
                    }
            }
```

Of course we could write it down more elegantly if needed

Comment (simone3991) [reply src/Clockify/ClockifyHandler.cs:131]: I'm also thinking that now we've complicated things by introducing KV, paying up the cost of token validation every time it's damn stupid. We could - at this stage - accept any non-null token, so no call to the KV is needed. If we catch an unauthorized exception during any dialog due to the token, we flag the user (via a `userProfile` field, for example. a `tokenExpired`) and at the next iteration we prompt them for a new valid one.

This way we reduce time and costs of communicating to the KV one time uselessly at every message

**Review by simone3991 — CHANGES_REQUESTED** [review]
I think we can improve the KV rollout phase, making it shorter

File: src/Clockify/ClockifySetupDialog.cs (hunk 1 above)
Comment (khaelys) [reply src/Clockify/ClockifySetupDialog.cs:83]: We can, but each user should re-enter the token, however, in this case line 83 can be erased

**Comment by khaelys** [comment]
As a reminder.

I noticed that in the codebase when a token is not found on TokenRepository read, a null value is returned. At first, it seemed a good API, but the entire codebase is cluttered with "check if missing and throw exception" because a "token not found" makes impossible all the operations.

I suggest throwing an exception directly in the repository.

**Comment by khaelys** [comment]
As a reminder.

I noticed that in the codebase when a token is not found on TokenRepository read, a null value is returned. At first, it seemed a good API, but the entire codebase is cluttered with "check if missing and throw exception" because a "token not found" makes impossible all the operations.

I suggest throwing an exception directly in the repository.

File: src/Clockify/ClockifyHandler.cs (hunk 2 above)
Comment (khaelys) [reply src/Clockify/ClockifyHandler.cs:130]: I prefer classic if/else, because one of the cases will be removed.

**Comment by simone3991** [comment]
> As a reminder.
> 
> I noticed that in the codebase when a token is not found on TokenRepository read, a null value is returned. At first, it seemed a good API, but the entire codebase is cluttered with "check if missing and throw exception" because a "token not found" makes impossible all the operations.
> 
> I suggest throwing an exception directly in the repository.

did you create an issue for this?

File: src/Remind/GenericRemindService.cs [hunk 3]
```diff
@@ -38,7 +38,7 @@ public async Task<string> SendReminderAsync(IBotFrameworkHttpAdapter adapter)
             List<UserProfile> userProfiles = await _userProfilesProvider.GetUserProfilesAsync();
 
             List<UserProfile> userToRemind = userProfiles
-                .Where(u => u.ClockifyToken != null && u.ConversationReference != null)
+                .Where(u => u.ClockifyTokenId != null && u.ConversationReference != null)
```
Comment (simone3991) [inline src/Remind/GenericRemindService.cs:41]: this is annoying, as we're likely to skip a few reminder for everyone.
Can we avoid it as a temporary solution (to be removed once we get rid of the field entirely)?

Comment (simone3991) [reply src/Remind/GenericRemindService.cs:41]: I checked, we don't really need the token, it was a smart move to avoid users who never signed in
So we can change it to

`.Where(u => (u.ClockifyTokenId ?? u.ClockifyToken) != null && u.ConversationReference != null)`

**Review by simone3991 — APPROVED** [review]
